domain/secret/state/state.go

```diff
@@ -3578,3 +3632,138 @@ func (st State) GetSecretsRevisionExpiryChanges(
+       JOIN secret_revision sr ON sr.secret_id = smo.secret_id
+       LEFT JOIN secret_revision_obsolete sro ON sro.revision_uuid = sr.uuid
+WHERE  sm.auto_prune = true AND sro.obsolete = true`
+		stmt, err := st.Prepare(q, secretExternalRevision{})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		var result secretExternalRevisions
+		err = runner.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+			err := tx.Query(ctx, stmt).GetAll(&result)
+			if errors.Is(err, sqlair.ErrNoRows) {
+				// It's ok, the revision probably has already been pruned.
+				return nil
+			}
+			return errors.Trace(err)
+		})
+		if err != nil {
+			return nil, errors.Trace(domain.CoerceError(err))
+		}
+		revisionUUIDs := make([]string, len(result))
+		for i, d := range result {
+			revisionUUIDs[i] = d.RevisionID
+		}
+		return revisionUUIDs, nil
+	}
+	return "secret_revision_obsolete", queryFunc
+}
+
+// GetObsoleteUserSecretRevisionReadyToPrune returns the specified user secret revision with secret ID if it is ready to prune.
+func (st State) GetObsoleteUserSecretRevisionsReadyToPrune(ctx context.Context, revisionIDs ...string) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	q := `
+SELECT (sr.revision, sr.secret_id) AS (&obsoleteRevisionRow.*)
+FROM   secret_model_owner smo
+       JOIN secret_metadata sm ON sm.secret_id = smo.secret_id
+       JOIN secret_revision sr ON sr.secret_id = smo.secret_id
+       LEFT JOIN secret_revision_obsolete sro ON sro.revision_uuid = sr.uuid
+WHERE  sm.auto_prune = true AND sro.obsolete = true`
+	var input []any
+	if len(revisionIDs) > 0 {
+		input = append(input, revisionUUIDs(revisionIDs))
+		q += " AND sr.uuid IN ($revisionUUIDs[:])"
+	}
+	stmt, err := st.Prepare(q, append(input, obsoleteRevisionRow{})...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var result obsoleteRevisionRows
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, input...).GetAll(&result)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// It's ok, the revision probably has already been pruned.
+			return nil
+		}
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return nil, errors.Trace(domain.CoerceError(err))
+	}
+	return result.toRevIDs(), nil
+}
+
+// InitialWatchStatementForUserSecretsToPrune returns the initial watch statement and the table name for watching user secrets with auto prune turned on.
+func (st State) InitialWatchStatementForUserSecretsToPrune() (string, eventsource.NamespaceQuery) {
+	queryFunc := func(ctx context.Context, runner coredatabase.TxnRunner) ([]string, error) {
+		result, err := st.getUserSecretRevisionsToPrune(ctx, runner)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		secretIDs := set.NewStrings()
+		for _, d := range result {
+			secretIDs.Add(d.SecretID)
+		}
+		return secretIDs.SortedValues(), nil
+	}
+	return "secret_metadata", queryFunc
```

Comment:
I think this use case is one where we need to consider using a custom change log event. This would remove the need for 2 separate watchers on different tables, both using the same SQL to query the secrets to act on. We can land this as is and consider refactoring.